stub/src/main/java/io/grpc/stub/annotations/RpcMethod.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2018 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.stub.annotations;
+
+import io.grpc.MethodDescriptor.MethodType;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotates a method descriptor method to provide metadata for annotation processing.
+ */
+@Retention(RetentionPolicy.CLASS)
+@Target(ElementType.METHOD)
+public @interface RpcMethod {
```

Comment:
The usage of this is not super clear to me.  For example, there are several other fields in MethodDescriptor that weren't included (like schema descriptor).  It needs to be clear that this is a limitted subset, and the source of truth is the getter itself.

Reply:
When using a compile time annotation processor you only have access to limited metadata.  Method names, class names, class variable names, parameter types, return types, and annotations but the class isn't loadable so you can't execute the code.  This annotation won't be available at runtime (Retention = CLASS) but it is available for annotation processing.  They are both sources of truth but each is provided for a different (and in fact, mutually exclusive) use case

Reply:
What I am worried about is misinterpretted intents.  There is very little documentation in the class about how it *should* be used.   Can anyone outside of grpc use this?  Could their annotation processor generate their own MethodDescriptors from this?  

If this really is only for annotation processors, why is MethodType included?  Where is there an example of how this will be used?

Reply:
Yes, anyone outside of grpc could use this (but only at compile time, not runtime).

Method type is the only field that is truly necessary (the others can be derived, albeit very crudely) for the annotation processor since there is no way today of distinguishing between client streaming and bidi streaming methods using the metadata available to an annotation processor.  Unary and server streaming can be distinguished by looking at the signatures in the BlockingStub class.

An example of where this could be useful is if you generated the implementation of the ImplBase to invoke some piece of common code you had with separate methods for each method type.

Reply:
Maybe something like: ?

> Annotates a method descriptor method to provide metadata for annotation processing. This can be used by annotation processors to discover gRPC information, as the MethodDescriptors are not available when the annotation processor runs.